### PR TITLE
zynaddsubfx: 3.0.5 → 3.0.6

### PR DESCRIPTION
**Pull request by kira-bruneau**
###### Description of changes

Update to the latest version: https://github.com/zynaddsubfx/zynaddsubfx/releases/tag/3.0.6

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kira-bruneau** [comment]
Result of `nixpkgs-review pr 185159` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>zyn-fusion</li>
    <li>zynaddsubfx-fltk</li>
    <li>zynaddsubfx-ntk</li>
  </ul>
</details>

File: pkgs/applications/audio/zynaddsubfx/default.nix [hunk 1]
```diff
@@ -15,7 +15,7 @@
 , zlib
 
   # Optional dependencies
-, alsaSupport ? true
+, alsaSupport ? stdenv.hostPlatform.isLinux
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zynaddsubfx/default.nix:18]:
```suggestion
, alsaSupport ? stdenv.isLinux
```

File: pkgs/applications/audio/zynaddsubfx/default.nix [hunk 2]
```diff
@@ -27,11 +27,13 @@
 , ossSupport ? true
 , portaudioSupport ? true
 , portaudio
+, sndioSupport ? stdenv.hostPlatform.isOpenBSD
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zynaddsubfx/default.nix:30]:
```suggestion
, sndioSupport ? stdenv.isOpenBSD
```

File: pkgs/applications/audio/zynaddsubfx/default.nix [hunk 3]
```diff
@@ -50,20 +53,24 @@ let
...
-    sha256 = "1vh1gszgjxwn8m32rk5222z1j2cnjax0bqpag7b47v6i36p2q4x8";
+    rev = "refs/tags/${version}";
     fetchSubmodules = true;
+    sha256 = "sha256-0siAx141DZx39facXWmKbsi0rHBNpobApTdey07EcXg=";
   };
 
+  outputs = [ "out" "doc" ];
+
   postPatch = ''
+    patchShebangs .
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zynaddsubfx/default.nix:73]: Please be more specific

File: pkgs/applications/audio/zynaddsubfx/default.nix [hunk 4]
```diff
@@ -87,29 +95,37 @@ in stdenv.mkDerivation rec {
...
     ctest --output-on-failure -E '^${lib.concatStringsSep "|" disabledTests}$'
     runHook postCheck
   '';
 
+  # Use Zyn-Fusion logo for zest build
+  # Manually derived from https://github.com/mruby-zest/mruby-zest-build/blob/3.0.6/src/mruby-zest/example/ZynLogo.qml#L65-L97
+  postInstall = lib.optionalString (guiModule == "zest") ''
+    rm -r "$out/share/pixmaps"
+    mkdir -p "$out/share/icons/hicolor/scalable/apps"
+    cp ${./ZynLogo.svg} "$out/share/icons/hicolor/scalable/apps/zynaddsubfx.svg"
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zynaddsubfx/default.nix:120]: Please fetchurl the logo

File: pkgs/applications/audio/zynaddsubfx/default.nix [hunk 5]
```diff
@@ -123,8 +139,9 @@ in stdenv.mkDerivation rec {
       then "https://zynaddsubfx.sourceforge.io/zyn-fusion.html"
       else "https://zynaddsubfx.sourceforge.io";
 
-    license = licenses.gpl2;
+    license = licenses.gpl2Plus;
     maintainers = with maintainers; [ goibhniu kira-bruneau ];
-    platforms = platforms.linux;
+    platforms = platforms.all;
+    broken = stdenv.isDarwin;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zynaddsubfx/default.nix:145]: Please add a comment why it is broken.

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -31821,19 +31821,20 @@ with pkgs;
...
 
-  zynaddsubfx-ntk = callPackage ../applications/audio/zynaddsubfx {
+  zynaddsubfx-ntk = zynaddsubfx.override {
     guiModule = "ntk";
   };
 
-  zyn-fusion = callPackage ../applications/audio/zynaddsubfx {
-    guiModule = "zest";
-  };
+  zyn-fusion = zynaddsubfx;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:31837]: Please move this to aliases.

File: pkgs/applications/audio/zynaddsubfx/default.nix (hunk 4 above)
Comment (kira-bruneau) [reply pkgs/applications/audio/zynaddsubfx/default.nix:120]: There's no SVG logo to fetch, I had to manually write the SVG by using the QML file that draws it in-app as a reference: https://github.com/mruby-zest/mruby-zest-build/blob/3.0.6/src/mruby-zest/example/ZynLogo.qml#L65-L97.

Comment (SuperSandro2000) [reply pkgs/applications/audio/zynaddsubfx/default.nix:120]: 😬 then we should add a comment that this question does not come up again.

Comment (kira-bruneau) [reply pkgs/applications/audio/zynaddsubfx/default.nix:120]: Oh yeah the comment I have there should probably be clearer about why I did that :sweat_smile:

File: pkgs/applications/audio/zynaddsubfx/default.nix (hunk 5 above)
Comment (kira-bruneau) [reply pkgs/applications/audio/zynaddsubfx/default.nix:145]: I don't remember exactly (I started this a while ago in my NUR repo), so instead I'm just going to try to get it to build.

File: pkgs/applications/audio/zynaddsubfx/default.nix [hunk 7]
```diff
@@ -123,8 +141,15 @@ in stdenv.mkDerivation rec {
...
     maintainers = with maintainers; [ goibhniu kira-bruneau ];
-    platforms = platforms.linux;
+    platforms = platforms.all;
+
+    # On macOS:
+    # - Tests don't compile (ld: unknown option: --no-as-needed)
+    # - ZynAddSubFX LV2 & VST plugin fail to compile (not setup to use ObjC version of pugl)
+    # - TTL generation crashes (`pointer being freed was not allocated`) for all VST plugins using AbstractFX
+    # - Zest UI fails to start on pulg_setup: Could not open display, aborting.
+    broken = stdenv.isDarwin;
```
Comment (kira-bruneau) [inline pkgs/applications/audio/zynaddsubfx/default.nix:153]: I went down a huge rabbit hole trying to get this working on Darwin. Turns out that the Darwin build is completely broken upstream, it looks like the latest version they have released as a binary is 3.0.3r20.

There's a bunch of things I fixed already, but it's not all the way there yet, and definitely shouldn't block the update (it wasn't building on Darwin before this change anyway).

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by kira-bruneau** [comment]
Result of `nixpkgs-review pr 185159` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>zynaddsubfx</li>
    <li>zynaddsubfx-fltk</li>
    <li>zynaddsubfx-ntk</li>
  </ul>
</details>

**Pull request merged**